vignettes/ggplot2-in-packages.Rmd

```diff
@@ -0,0 +1,241 @@
+
+```{r}
+#' @importFrom ggplot2 ggplot aes geom_bar coord_flip
+mpg_drv_summary <- function() {
+  ggplot(ggplot2::mpg) + 
+    geom_bar(aes(x = drv)) + 
+    coord_flip()
+}
+```
+
+```{r, include=FALSE}
+# make sure this function runs!
+mpg_drv_summary()
+```
+
+Even if you use many ggplot2 functions in your package, it is unwise to use ggplot2 in `Depends` or import the entire package into your `NAMESPACE`. Using ggplot2 in `Depends` will attach ggplot2 when your package is attached, which includes when your package is tested. This makes it difficult to ensure that others can use the functions in your package without attaching it (i.e., using `::`). Similarly, importing all 450 of ggplot2's exported objects into your namespace makes it difficult to separate the responsibility of your package and the responsibility of ggplot2, in addition to making it difficult for readers of your code to figure out where functions are coming from!
+
+## Using `aes()` and `vars()` in a package function
+
+To create any graphic using ggplot2 you will probably need to use `aes()` at least once. If your graphic uses facets, you might be using `vars()` to refer to columns in the plot/layer data. Both of these functions use non-standard evaluation, so if you try to use them in a function within a package they will result in a CMD check note:
+
+```{r}
+#' @importFrom ggplot2 ggplot aes geom_bar coord_flip
+mpg_drv_summary <- function() {
+  ggplot(ggplot2::mpg) + 
+    geom_bar(aes(x = drv)) + 
+    coord_flip()
+}
+```
+
+```
+N  checking R code for possible problems (2.7s)
+   mpg_drv_summary: no visible binding for global variable ‘drv’
+   Undefined global functions or variables:
+     drv
+```
+
+There are three situations in which you will encounter this problem:
+
+- You already know the column name or expression in advance.
+- You have the column name as a character vector.
+- The user specifies the column name or expression, and you want your function to use the same kind of non-standard evaluation used by `aes()` and `vars()`.
+
+If you already know the mapping in advance (like the above example) you should use the `.data` pronoun from [rlang](https://cran.r-project.org/package=rlang) to make it explicit that you are referring to the `drv` in the data and not some other variable named `drv` (which may or may not exist elsewhere).
+
+```{r}
+#' @importFrom rlang .data
+mpg_drv_summary <- function() {
+  ggplot(mpg) + 
+    geom_bar(aes(x = .data$drv)) + 
+    coord_flip()
+}
+```
+
+If you have the column name as a character vector (e.g., `col = "drv"`), use `.data[[col]]`:
+
+```{r}
+#' @importFrom rlang .data
+col_summary <- function(data, col) {
+  ggplot(mpg) + 
+    geom_bar(aes(x = .data[[col]])) + 
+    coord_flip()
+}
+
+col_summary(mpg, "drv")
+```
+
+To use the same kind of non-standard evaluation used by `aes()` and `vars()`, use `{{ col }}` to pass the unevaluated expression the user typed in `col` to `aes()`.
```

Comment:
Perhaps:

If the column name is supplied by the user, you can also interpolate it with `{{ col }}`. This tidy eval operator captures the expression supplied by the user and forwards it to another quoting function such as `aes()` or `vars()`.